ktlint-reporter-sarif/src/main/kotlin/com/pinterest/ktlint/reporter/sarif/SarifReporter.kt

```diff
@@ -0,0 +1,97 @@
+import com.pinterest.ktlint.core.getKtlintVersion
+import io.github.detekt.sarif4k.ArtifactLocation
+import io.github.detekt.sarif4k.Level
+import io.github.detekt.sarif4k.Location
+import io.github.detekt.sarif4k.Message
+import io.github.detekt.sarif4k.PhysicalLocation
+import io.github.detekt.sarif4k.Region
+import io.github.detekt.sarif4k.Result
+import io.github.detekt.sarif4k.Run
+import io.github.detekt.sarif4k.SarifSchema210
+import io.github.detekt.sarif4k.SarifSerializer
+import io.github.detekt.sarif4k.Tool
+import io.github.detekt.sarif4k.ToolComponent
+import io.github.detekt.sarif4k.Version
+import java.io.File
+import java.io.PrintStream
+
+private const val SRCROOT = "%SRCROOT%"
+internal fun String.sanitize(): String =
+    this.replace(File.separatorChar, '/')
+        .let {
+            if (it.endsWith('/')) it else "$it/"
+        }
+
+public class SarifReporter(private val out: PrintStream) : Reporter {
+
+    private val results: MutableList<Result> = mutableListOf()
+    private var workingDirectory: File? = null
+
+    override fun beforeAll() {
+        workingDirectory = System.getProperty("user.home")?.let(::File)
+    }
+
+    override fun onLintError(file: String, err: LintError, corrected: Boolean) {
+        results.add(
+            Result(
+                ruleID = err.ruleId,
+                level = Level.Error,
+                locations = listOf(
+                    Location(
+                        physicalLocation = PhysicalLocation(
+                            region = Region(
+                                startLine = err.line.toLong(),
+                                startColumn = err.col.toLong()
+                            ),
+                            artifactLocation = workingDirectory?.let { workingDirectory ->
+                                ArtifactLocation(
+                                    uri = File(file).relativeTo(workingDirectory).path,
+                                    uriBaseID = SRCROOT
+
+                                )
+                            } ?: ArtifactLocation(
+                                uri = file
+                            )
+                        )
+                    )
+                ),
+                message = Message(text = err.detail)
+            )
+        )
+    }
+
+    override fun afterAll() {
+        val version = getKtlintVersion(SarifReporter::class.java)
+        val sarifSchema210 = SarifSchema210(
+            schema = "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
+            version = Version.The210,
+            runs = listOf(
+                Run(
+                    tool = Tool(
+                        driver = ToolComponent(
+                            downloadURI = "https://github.com/pinterest/ktlint/releases/tag/$version",
+                            fullName = "ktlint",
+                            guid = "7e96e861-6db7-44ea-a1e4-56bb4554ffe9",
```

Comment:
hm, where is this coming from?

Reply:
guid is a unique identifier for the component: https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317539. I generated a random one for the time being.

It is optional, I can remove this.